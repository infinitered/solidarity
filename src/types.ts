import { GluegunRunContext } from 'gluegun'
export const solidarity = {
  binaryExists: require('./extensions/functions/binaryExists'),
  getSolidaritySettings: require('./extensions/functions/getSolidaritySettings'),
  setSolidaritySettings: require('./extensions/functions/setSolidaritySettings'),
  checkRequirement: require('./extensions/functions/checkRequirement'),
  updateRequirement: require('./extensions/functions/updateRequirement'),
  skipRule: require('./extensions/functions/skipRule'),
  checkENV: require('./extensions/functions/checkENV'),
  checkCLI: require('./extensions/functions/checkCLI'),
  checkDir: require('./extensions/functions/checkDir'),
  checkFile: require('./extensions/functions/checkFile'),
  updateVersions: require('./extensions/functions/updateVersions'),
  removeNonVersionCharacters: require('./extensions/functions/removeNonVersionCharacters'),
  getVersion: require('./extensions/functions/getVersion'),
  getLineWithVersion: require('./extensions/functions/getLineWithVersion')
}

export interface SolidarityPlugin {
  name: string
  description: string
  snapshot: string | SnapshotType
}

export interface SolidarityRunContext extends GluegunRunContext {
  solidarity: typeof solidarity
  _pluginsList: Array<SolidarityPlugin & {templateDirectory: string}>
  addPlugin: (config: SolidarityPlugin) => void
  printSeparator: () => void
  outputMode: SolidarityOutputMode
}

export type SnapshotType = (context: SolidarityRunContext) => Promise<void>

export interface SolidarityRule {
  rule: 'cli' | 'env' | 'dir' | 'file'
  binary?: string
  semver?: string
  version?: string
  line?: string | number
  variable?: string
  location?: string
  error?: string
  platform?: string | string[]
  matchIndex?: number
}

export type SolidarityRequirement = SolidarityRule[]

export interface SolidaritySettings {
  requirements: object
  config: object
}

export enum SolidarityOutputMode {
  MODERATE,
  VERBOSE,
  SILENT
}
