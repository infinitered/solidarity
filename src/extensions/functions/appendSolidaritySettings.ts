import { keys } from 'ramda'

module.exports = (solidaritySettings, newRequirement) => {
  const newRequirementKey = keys(newRequirement)
  const existingRequirementRules = solidaritySettings.requirements[newRequirementKey] || []

  return {
    ...solidaritySettings,
    requirements: {
      ...solidaritySettings.requirements,
      [newRequirementKey]: [
        ...existingRequirementRules,
        ...newRequirement[newRequirementKey]
      ]
    }
  }
}
