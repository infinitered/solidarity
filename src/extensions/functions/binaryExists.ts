import { SolidarityRunContext } from '../../types'
module.exports = (binary: string, context: SolidarityRunContext): boolean => {
  const { system } = context

  // Check if binary exists
  try {
    system.which(binary)
    return true
  } catch (_e) {
    return false
  }
}
