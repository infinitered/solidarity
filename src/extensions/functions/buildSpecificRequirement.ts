import { SolidarityRunContext, SolidaritySettings } from '../../types'
import { keys } from 'ramda'

namespace buildSpecificRequirement {
  const requiredInputQuestion = async ({ name, message, prompt }) => {
    const result = await prompt.ask({
      name,
      type: 'input',
      message,
    })

    if (!result[name]) return Promise.reject('An input is required')
    return result
  }

  const resolveParameters = async ({ parameters, prompt, ruleHandlers }) => {
    const { first, second } = parameters
    if (second) {
      return Promise.resolve({ whatRule: second })
    }

    const message = `What's the ${first} ${ruleHandlers[first].key} you'd like to add a rule for?`

    return requiredInputQuestion({
      name: 'whatRule',
      message,
      prompt,
    })
  }

  const getRequirementNames = (solidaritySettings: SolidaritySettings): Array<string> =>
    keys(solidaritySettings.requirements)

  const chooseRequirement = async (prompt, solidaritySettings: SolidaritySettings): Promise<String> => {
    const shouldMakeNewRequirement = await prompt.ask({
      name: 'makeNewRequirement',
      type: 'confirm',
      message: 'Would you like to create a new requirement set?',
    })

    let requirementName

    if (shouldMakeNewRequirement.makeNewRequirement) {
      const answer = await requiredInputQuestion({
        name: 'newRequirement',
        message: 'What would you like to call this new requirement?',
        prompt,
      }).catch(error => {
        return Promise.reject(error)
      })
      requirementName = answer.newRequirement
    } else {
      const requirementOptions = getRequirementNames(solidaritySettings)
      const answer = await prompt.ask({
        name: 'selectedRequirement',
        message: 'Which of the above technology snapshots will you use for this project?',
        type: 'selet',
        choices: requirementOptions,
      })
      requirementName = answer.selectedRequirement
    }

    return requirementName
  }

  const constructRequirment = async context => {
    const { parameters, prompt, solidarity } = context
    const { getSolidaritySettings, ruleHandlers } = solidarity
    const solidaritySettings = await getSolidaritySettings(context)

    const userAnswer = await prompt.ask({
      name: 'addNewRule',
      type: 'confirm',
      message: `Would you like to add the ${parameters.first} '${parameters.second}' to your Solidarity file?`,
    })

    if (userAnswer.addNewRule) {
      // maybe ask about setting up the new rule w/ a specific version?
      const requirementName = await chooseRequirement(prompt, solidaritySettings)
      return ruleHandlers[parameters.first].callback(context, requirementName)
    } else {
      return Promise.reject('Rule not added.')
    }
  }

  export const run = async (context: SolidarityRunContext) => {
    const { parameters, prompt, solidarity } = context
    const { first } = parameters
    const { ruleHandlers } = solidarity
    const resolvedParam = await resolveParameters({ parameters, prompt, ruleHandlers }).catch(() => {
      return Promise.reject('Missing required parameters.')
    })

    return constructRequirment({
      ...context,
      parameters: {
        ...parameters,
        first,
        second: resolvedParam.whatRule,
      },
    })
  }
}

module.exports = buildSpecificRequirement.run
