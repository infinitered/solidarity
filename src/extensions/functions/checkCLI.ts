import { SolidarityRunContext, SolidarityRule } from '../../types'
module.exports = async (rule: SolidarityRule, context: SolidarityRunContext): Promise<string> => {
  const { semver, solidarity } = context
  const binaryExists = require('./binaryExists')

  // First check for binary
  if (!binaryExists(rule.binary, context)) {
    return `Binary '${rule.binary}' not found`
  }

  // Is there a semver rule?
  if (rule.semver) {
    // ensure we have valid rule input
    if (!semver.validRange(rule.semver)) return `Invalid semver rule ${rule.semver}`

    const binaryVersion = await solidarity.getVersion(rule, context)
    // pad zeros for any non-semver version systems (rules still work)
    let binarySemantic = binaryVersion
    while (binarySemantic.split('.').length < 3) { binarySemantic += '.0' }

    // I can't get no satisfaction
    if (!semver.satisfies(binarySemantic, rule.semver)) {
      return `${rule.binary}: you have '${binaryVersion}', but the project requires '${rule.semver}'`
    }
  }
}
