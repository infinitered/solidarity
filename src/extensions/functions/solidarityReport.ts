import { SolidarityRunContext, SolidarityReportResults, CLIReportConfig } from '../../types'
import { helpers } from 'envinfo'

export const createReport = (context: SolidarityRunContext): SolidarityReportResults => {
  const { print, system } = context
  const { colors } = print

  return {
    basicInfo: [['System Basics', 'Value'], ['OS', helpers.getOperatingSystemInfo()], ['CPU', helpers.getCPUInfo()]],
    cliRules: [['Binary', 'Location', 'Version', 'Desired']],
    envRules: [['Environment Var', 'Value']],
    filesystemRules: [['Location', 'Type', 'Exists']],
    shellRules: [['Command', 'Pattern', 'Matches']],
    customRules: [],
    // helper for adding CLI rules
    addCLI: function (cliReportConfig: CLIReportConfig) {
      const desired = cliReportConfig.desired ? cliReportConfig.desired : colors.green('*ANY*')
      let location
      try {
        location = system.which(cliReportConfig.binary)
      } catch (_e) {
        location = colors.red('*MISSING*')
      }
      this.cliRules.push([cliReportConfig.binary, location, cliReportConfig.version, desired])
    },
  }
}
