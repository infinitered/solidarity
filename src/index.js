const { build } = require('gluegun')
const {dependenciesMet} = require('gluegun/sniff')
if(dependenciesMet==1){
    module.exports = async () => {
      // setup the runtime
      build()
        .brand('solidarity')
        .src(__dirname)
        // It might also be worth checking global
        // `npm config get prefix` returns '/usr/local'
        // global npm is stored /usr/local/lib/node_modules
        .plugins('./node_modules', { matching: 'solidarity-*', hidden: true })
        // for testing - force load a plugin
        // .plugin('../solidarity-react-native')
        .create()
        .run()
    }
}
