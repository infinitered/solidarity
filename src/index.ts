import { build } from 'gluegun'

module.exports = async (): Promise<void> => {
  // setup the runtime
  build()
    .brand('solidarity')
    .src(__dirname)
    // local installs
    .plugins('./node_modules', { matching: 'solidarity-*', hidden: true })
    // global installs
    .plugins('/usr/local/lib/node_modules', { matching: 'solidarity-*', hidden: true })
    // for testing - force load a local plugin
    // .plugin('../solidarity-react-native')
    .create()
    .run()
}

export * from './types'
