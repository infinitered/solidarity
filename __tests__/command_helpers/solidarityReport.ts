import { SolidarityReportResults } from '../../src/types'
import { createReport } from '../../src/extensions/functions/solidarityReport'
const context = require('mockContext')

describe('solidarityReport structure', () => {
  test('the basic function generates the Result object', () => {
    let report = createReport(context)
    // Check everything but system stuff against snapshots
    expect(report.addCLI).toMatchSnapshot()
    expect(report.cliRules).toMatchSnapshot()
    expect(report.customRules).toMatchSnapshot()
    expect(report.envRules).toMatchSnapshot()
    expect(report.filesystemRules).toMatchSnapshot()
    expect(report.shellRules).toMatchSnapshot()
  })
})
