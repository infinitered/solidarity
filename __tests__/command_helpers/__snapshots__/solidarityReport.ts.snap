// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`solidarityReport structure the basic function generates the Result object 1`] = `[Function]`;

exports[`solidarityReport structure the basic function generates the Result object 2`] = `
Array [
  Array [
    "Binary",
    "Location",
    "Version",
    "Desired",
  ],
]
`;

exports[`solidarityReport structure the basic function generates the Result object 3`] = `Array []`;

exports[`solidarityReport structure the basic function generates the Result object 4`] = `
Array [
  Array [
    "Environment Var",
    "Value",
  ],
]
`;

exports[`solidarityReport structure the basic function generates the Result object 5`] = `
Array [
  Array [
    "Location",
    "Type",
    "Exists",
  ],
]
`;

exports[`solidarityReport structure the basic function generates the Result object 6`] = `
Array [
  Array [
    "Command",
    "Pattern",
    "Matches",
  ],
]
`;
